third_party/tcplp/CMakeLists.txt

```diff
@@ -43,48 +43,65 @@ set(src_tcplp
     lib/lbuf.c
 )
 
-set(tcplp_static_target "tcplp")
 
 string(REPLACE "-Wsign-compare" "" CMAKE_C_FLAGS "${CMAKE_C_FLAGS}")
 
 string(REPLACE "-Wconversion" "" CMAKE_C_FLAGS "${CMAKE_C_FLAGS}")
 
 string(REPLACE "-Wunused-parameter" "" CMAKE_C_FLAGS "${CMAKE_C_FLAGS}")
 
-add_library(${tcplp_static_target} STATIC ${src_tcplp})
-target_compile_options(${tcplp_static_target}
-    PRIVATE
-        "-Wno-sign-compare"
-        "-Wno-unused-parameter"
-)
-set_target_properties(${tcplp_static_target} PROPERTIES OUTPUT_NAME tcplp)
-target_include_directories(${tcplp_static_target}
-    PUBLIC
-        ${CMAKE_CURRENT_SOURCE_DIR}/bsdtcp
-        ${CMAKE_CURRENT_SOURCE_DIR}/lib
-    PRIVATE
-        ${OT_PUBLIC_INCLUDES}
-)
-
-target_link_libraries(${tcplp_static_target}
-    PRIVATE
-        ot-config
-)
+if(OT_FTD)
+    add_library(tcplp-ftd STATIC ${src_tcplp})
+    target_compile_options(tcplp-ftd
+        PRIVATE
+            "-Wno-sign-compare"
+            "-Wno-unused-parameter"
+    )
+    set_target_properties(tcplp-ftd PROPERTIES OUTPUT_NAME tcplp-ftd)
+    target_include_directories(tcplp-ftd
+        PUBLIC
+            ${CMAKE_CURRENT_SOURCE_DIR}/bsdtcp
+            ${CMAKE_CURRENT_SOURCE_DIR}/lib
+        PRIVATE
+            ${OT_PUBLIC_INCLUDES}
+    )
 
-# TCPlp calls functions that are defined by the core OpenThread (like
-# "otMessageWrite()"), so we need to add the core library (FTD or MTD, as
-# appropriate) as a link dependency.
+    target_link_libraries(tcplp-ftd
+        PRIVATE
+            ot-config
+    )
 
-if(OT_FTD)
-    target_link_libraries(${tcplp_static_target}
+    target_link_libraries(tcplp-ftd
         PRIVATE
             openthread-ftd
     )
+
 endif()
 
 if(OT_MTD)
-    target_link_libraries(${tcplp_static_target}
+    add_library(tcplp-mtd STATIC ${src_tcplp})
+    target_compile_options(tcplp-mtd
+        PRIVATE
+            "-Wno-sign-compare"
+            "-Wno-unused-parameter"
+    )
+    set_target_properties(tcplp-mtd PROPERTIES OUTPUT_NAME tcplp-mtd)
+    target_include_directories(tcplp-mtd
+        PUBLIC
+            ${CMAKE_CURRENT_SOURCE_DIR}/bsdtcp
+            ${CMAKE_CURRENT_SOURCE_DIR}/lib
+        PRIVATE
+            ${OT_PUBLIC_INCLUDES}
+    )
+
+    target_link_libraries(tcplp-mtd
+        PRIVATE
+            ot-config
+    )
+
+    target_link_libraries(tcplp-ftd
```

Comment:
```suggestion
    target_link_libraries(tcplp-mtd
```

Reply:
Ah! Thanks. I changed this on https://github.com/openthread/openthread/pull/8163 as well to make sure this is still fixing the linker error.

Reply:
Turned out this creates a new linker error: 
With this we now get a new error:https://github.com/openthread/openthread/actions/runs/3084942096/jobs/4987679455
(seems like when linking `tcplp` does not see certain OT lib methods).

It is curios why we see the error only for `ot-test-heap-array` and not any other tests? (I cannot telll what may be different/special about this one test.

@samkumar  do you happen to have any suggestion/idea on this?

Reply:
Seems like including `openthread-ftd` twice in unit test cmake file addresses the above error.